keepalived-vip/controller.go

```diff
@@ -0,0 +1,253 @@
+	reloadRateLimiter util.RateLimiter
+	keepalived        *keepalived
+	reloadLock        *sync.Mutex
+}
+
+// getEndpoints returns a list of <endpoint ip>:<port> for a given service/target port combination.
+func (ipvsc *ipvsControllerController) getEndpoints(
+	s *api.Service, servicePort *api.ServicePort) (endpoints []service) {
+	ep, err := ipvsc.epLister.GetServiceEndpoints(s)
+	if err != nil {
+		return
+	}
+
+	// The intent here is to create a union of all subsets that match a targetPort.
+	// We know the endpoint already matches the service, so all pod ips that have
+	// the target port are capable of service traffic for it.
+	for _, ss := range ep.Subsets {
+		for _, epPort := range ss.Ports {
+			var targetPort int
+			switch servicePort.TargetPort.Type {
+			case intstr.Int:
+				if epPort.Port == servicePort.TargetPort.IntValue() {
+					targetPort = epPort.Port
+				}
+			case intstr.String:
+				if epPort.Name == servicePort.TargetPort.StrVal {
+					targetPort = epPort.Port
+				}
+			}
+			if targetPort == 0 {
+				continue
+			}
+			for _, epAddress := range ss.Addresses {
+				endpoints = append(endpoints, service{Ip: epAddress.IP, Port: targetPort})
+			}
+		}
+	}
+	return
+}
+
+// getVIPs returns list of virtual IPs to expose.
+func (ipvsc *ipvsControllerController) getVIPs() []string {
+	vips := []string{}
+
+	services, _ := ipvsc.svcLister.List()
+	for _, s := range services.Items {
+		if externalIP, ok := s.GetAnnotations()[ipvsPublicVIP]; ok {
+			glog.Infof("Service %v contains public-vip annotation", s.Name)
+			vips = append(vips, externalIP)
+		}
+	}
+
+	return vips
+}
+
+// getServices returns a list of services and their endpoints.
+func (ipvsc *ipvsControllerController) getServices() []vip {
+	svcs := []vip{}
+
+	services, _ := ipvsc.svcLister.List()
+	for _, s := range services.Items {
+		if externalIP, ok := s.GetAnnotations()[ipvsPublicVIP]; ok {
+			for _, servicePort := range s.Spec.Ports {
+				ep := ipvsc.getEndpoints(&s, &servicePort)
+				if len(ep) == 0 {
+					glog.Infof("No endpoints found for service %v, port %+v", s.Name, servicePort)
+					continue
+				}
+
+				svcs = append(svcs, vip{
+					Name:     fmt.Sprintf("%v/%v", s.Namespace, s.Name),
+					Ip:       externalIP,
+					Port:     servicePort.Port,
+					Backends: ep,
+					Protocol: fmt.Sprintf("%v", servicePort.Protocol),
+				})
+				glog.Infof("Found service: %v", s.Name)
+			}
+		}
+	}
+
+	return svcs
+}
+
+// sync all services with the loadbalancer.
+func (ipvsc *ipvsControllerController) sync() error {
+	ipvsc.reloadLock.Lock()
+	defer ipvsc.reloadLock.Unlock()
+
+	if !ipvsc.epController.HasSynced() || !ipvsc.svcController.HasSynced() {
+		time.Sleep(100 * time.Millisecond)
+		return errDeferredSync
+	}
+
+	err := ipvsc.keepalived.WriteCfg(ipvsc.getServices())
+	if err != nil {
+		return err
+	}
+
+	err = ipvsc.keepalived.Reload()
+	if err != nil {
+		return err
+	}
+
+	ipvsc.reloadRateLimiter.Accept()
```

Comment:
This takes tokens out of the tokenbucket and blocks if none are available, so if you're calling sync from multiple threads wouldn't you want them to block at the head of this function instead?